### DISCUSSION TITLE
PWA / WASM

### DISCUSSION THREAD

**Discussion by alensiljak** [ananthakumaran/paisa#30] [General]
Hi! I'd like a raise a question/idea about a multi-plaform technological solution for a ledger GUI / analytics. 
As a brief background, I have a goal similar to what Paisa is doing - to have a multi-platform (i.e. web browser) GUI for the reports and insights into my Ledger book.
So, what I've been thinking of, and (very) slowly implementing with [Ledger-rs](https://github.com/ledger-rs/) organization and [Cashier](https://cashier.alensiljak.eu.org/) PWA, is to use a PWA as the UI. It is offline and client-only (not -first). There is a server component (Cashier Sync), which is a simple proxy to ledger-cli. 
Conceptually, this is what Paisa is doing. I'm using an in-browser database but that is a relatively minor difference. And, with SQLite-Wasm, this could be a non-problem, as well.
What remains in the logic might be published as a WASM component, making it fully in-browser solution. Or, with WASI, it could be a server-side component.

Anyways, just wanted to see if you have been thinking about something similar and/or if you'd be willing to participate in efforts to get there. The benefits are obvious (I think) and the amount of work is not that much bigger than implementing anything else. Plus it's a great learning experience.
I hope you get the gist of my idea from this scribble. :)

**Comment by ananthakumaran** [comment]
I did spend some time trying to compile the ledger using Emscripten but gave up quickly. I don't think I have the bandwidth to re-implement the ledger, but if there is a wasm based ledger available, I would definitely consider going PWA route with Paisa (few unknowns need to be sorted, specially PWS won't allow cors requests, which is necessary for price fetching). For now, I am happy with the desktop application and would mostly focus on the UI side.
